tokio/src/util/wake_list.rs

```diff
@@ -37,12 +37,35 @@ impl WakeList {
     }
 
     pub(crate) fn wake_all(&mut self) {
-        assert!(self.curr <= NUM_WAKERS);
-        while self.curr > 0 {
-            self.curr -= 1;
-            // SAFETY: The first `curr` elements of `WakeList` are initialized, so by decrementing
-            // `curr`, we can take ownership of the last item.
-            let waker = unsafe { ptr::read(self.inner[self.curr].as_mut_ptr()) };
+        struct DropGuard {
+            start: *mut Waker,
+            end: *mut Waker,
+        }
+
+        impl Drop for DropGuard {
+            fn drop(&mut self) {
+                // SAFETY: Both pointers are part of the same object, with `start <= end`.
+                let len = unsafe { self.end.offset_from(self.start) } as usize;
+                let slice = ptr::slice_from_raw_parts_mut(self.start, len);
+                // SAFETY: All elements in `start..len` are initialized, so we can drop them.
+                unsafe { ptr::drop_in_place(slice) };
+            }
+        }
+
+        debug_assert!(self.curr <= NUM_WAKERS);
+
+        let mut guard = {
+            let len = mem::take(&mut self.curr);
```

Comment:
I think using `mem::take` to set it to zero is confusing. Instead, I would do so explicitly, and have a comment saying that we set it to zero and thereby transfer ownership to the drop guard.